lib/blocks.php

```diff
@@ -93,3 +159,42 @@ function do_blocks( $content ) {
 	return $content_after_blocks;
 }
 add_filter( 'the_content', 'do_blocks', 9 ); // BEFORE do_shortcode() and wpautop().
+
+/**
+ * Given a string, returns content normalized with automatic paragraphs applied
+ * to text not identified as a block. Since this executes the block parser, it
+ * should not be used in a performance-critical flow such as content display.
+ * Block content will not have automatic paragraphs applied.
+ *
+ * @param  string $content Original content.
+ * @return string          Content formatted with automatic paragraphs applied
+ *                         to unknown blocks.
+ */
+function wpautop_block_content( $content ) {
+	$blocks = gutenberg_parse_blocks( $content );
+	foreach ( $blocks as $i => $block ) {
+		if ( isset( $block['blockName'] ) ) {
+			continue;
+		}
+
+		$content = $block['rawContent'];
+
+		// wpautop will trim leading whitespace and return whitespace-only text
+		// as an empty string. Preserve to apply leading whitespace as prefix.
+		preg_match( '/^(\s+)/', $content, $prefix_match );
+		$prefix = empty( $prefix_match ) ? '' : $prefix_match[0];
+
+		$content = $prefix . wpautop( $content, false );
+
+		// To normalize as text where wpautop would not be applied, restore
+		// double newline to wpautop'd text if not at the end of content.
+		$is_last_block = ( count( $blocks ) === $i + 1 );
+		if ( ! $is_last_block ) {
+			$content = str_replace( "</p>\n", "</p>\n\n", $content );
+		}
+
+		$blocks[ $i ]['rawContent'] = $content;
+	}
+
+	return gutenberg_serialize_blocks( $blocks );
```

Comment:
This is one place where I have concern about using HTML strings as Editable (cc: @iseulde @youknowriad @mtias).

If we could iterate over the text nodes themselves and know that they contain no HTML then it seems like we could guarantee behaviors with `autop`. In this mixture, however, it feels like we leave the door open to bugs. Maybe it's not too big of a deal. I think `<` is illegal inside attributes and tags, but not illegal inside comments.

Not sure also how `autop` works so please read my review with that in mind.

Reply:
I'm not quite sure I understand the concern here. With these changes and #2708, we're effectively turning off `wpautop` for blocks, so it shouldn't be the case that we need to worry about how rich text (via HTML or other structure) is treated.

Reply:
more of a general concern about existing `apply_filter()` code